### PR TITLE
#59 Fix PDF view broken on macOS 11

### LINKED ISSUE #59: PDF view broken on macOS 11
On macOS 11 Big Sur, the PDF view displays PDFs incorrectly.

### PR DESCRIPTION
**Pull request by thSoft**
Fixes #59.

### CODE REVIEW

**Comment by nittka** [comment]
Unfortunately, the change breaks rendering under Windows.
renderingWindows

Do you have an idea where to influence the transparency of the hyperlinks? I think, we had a similar issue in the past.

Also when navigating from source to score (F3), the score flickers - possibly because with each phase of highlighting the annotation (blending in and out), the image is painted again? A number of exceptions are flying around as well. But they should be dealt with separately.
[Edit] The exceptions are not caused by the change. They occur already on the current master branches. I guess, they are due to the updated Eclipse/Java setup (I guess you don't experience them!?)

(The error already occurs before updating the jpedal version.) I have not tested performance with big files, yet.

File: org.eclipse.ui.views.pdf/src/org/eclipse/ui/views/pdf/PdfViewPage.java [hunk 1]
```diff
@@ -54,6 +54,14 @@ public PdfViewPage(Composite parent, IFile file) throws PdfException {
 		super(parent, SWT.H_SCROLL | SWT.V_SCROLL);
 		pdfDisplay = new Composite(this, SWT.NONE);
 		pdfDisplay.setBackgroundMode(SWT.INHERIT_FORCE);
+		pdfDisplay.addPaintListener(new PaintListener() {
+			@Override
+			public void paintControl(PaintEvent event) {
+				if (image != null) {
+					event.gc.drawImage(image, 0, 0);
+				}
+		    }
```
Comment (nittka) [inline org.eclipse.ui.views.pdf/src/org/eclipse/ui/views/pdf/PdfViewPage.java:63]: tab instead of spaces

**Comment by nittka** [comment]
With respect to the broken score: it seems swt for windows has been fixed and we do not need the special style handling in PdfAnnotationHyperlink anymore - `SWT.TRANSPARENT | SWT.NO_BACKGROUND` now also works for windows. With this change the score looks OK, the flickering on navigation is gone. But not the exceptions and on my machine the link animation is not very smooth. I'll investigate the exceptions further.

**Comment by nittka** [comment]
From what I see, the paint event really needs to paint the image every time (even if it did not change). But if a hyperlink is highlighted/activated with F3, the paint method starts being called very often (until fadeout is over). However, on my machine this takes a loong time.
If more than 255 Hyperlinks are created, a call stack exception is thrown (even with cursor setting and mouseadapter being commented out). And even then the hyperlink animation cycle takes quite a long time.
I don't yet know, what causes this problem and what could be done about it.

**Comment by nittka** [comment]
It seems, my setup was messed up yesterday - maybe some heavy CPU usage causing all the side effects. Except for the rendering problem, the change for rendering the image seems fine. I suggest further replacing the PdfAnnotationHyperlink-code by 
```
package org.eclipse.ui.views.pdf;

import org.eclipse.swt.SWT;
import org.eclipse.swt.events.MouseAdapter;
import org.eclipse.swt.events.MouseEvent;
import org.eclipse.swt.events.MouseListener;
import org.eclipse.swt.widgets.Composite;
import org.eclipse.swt.widgets.Display;
import org.eclipse.util.TextEditorUtils;

public class PdfAnnotationHyperlink extends Composite {

	private static MouseListener LISTENER = new MouseAdapter() {
		@Override
		public void mouseDown(MouseEvent e) {
			Object source = e.getSource();
			if (source instanceof PdfAnnotationHyperlink) {
				PdfAnnotation annotation = ((PdfAnnotationHyperlink) source).annotation;
				if (annotation != null) {
					if (annotation.fileURI != null) {
						TextEditorUtils.revealPosition(annotation.fileURI, annotation.lineNumber,
								annotation.columnNumber, 1);
					} else if (annotation.file != null) {
						TextEditorUtils.revealPosition(annotation.file, annotation.lineNumber, annotation.columnNumber,
								1);
					}
				}
			}
		}
	};

	private PdfAnnotation annotation;

	public PdfAnnotationHyperlink(Composite parent, final PdfAnnotation annotation) {
		super(parent, SWT.TRANSPARENT | SWT.NO_BACKGROUND);
		this.annotation = annotation;
		setCursor(Display.getDefault().getSystemCursor(SWT.CURSOR_HAND));
		addMouseListener(LISTENER);
	}

}
```
* no need for a different style under Windows, anymore
* use the recommended cursor code (see java doc in the current implementation)
* prevent instantiating hundreds of mouse adapters

Two further issues need to be addressed. If you have a score with more than one page and you invoke a source to score navigation, the hyperlink is shown for some time. 
1. If during that time, you go to the next page, the hyperlink is still shown on the new page. Only if you change the page once more, it disappears.
2. If during that time, you close the score, a "widget dispose"-Exception is thrown in HyperlinkHighlightAnimator.run(PdfViewPage.java:882). Here not null and not disposed guard needs to be added.

Could you check whether you can reproduce them? At least the second problem should be not be Windows-specific.

**Comment by thSoft** [comment]
> 1. If during that time, you go to the next page, the hyperlink is still shown on the new page. Only if you change the page once more, it disappears.

I can't reproduce this.

> 2. If during that time, you close the score, a "widget dispose"-Exception is thrown in HyperlinkHighlightAnimator.run(PdfViewPage.java:882). Here not null and not disposed guard needs to be added.

I can reproduce this.

**Review by nittka — APPROVED** [review]
Either you add the disposed-guard within this PR or I'll do so while trying to resolve the Windows-specific hyperlink-non-removal.

**Comment by nittka** [comment]
Please check the last commit of branch issues/59_nittka (https://github.com/thSoft/eclipse-commons/tree/issues/59_nittka) for further changes I suggest in PdfViewPage. They deal with both the issues I raised above. If you are fine with adding this commit to your PR, everything should be ready for merging.

**Comment by thSoft** [comment]
Great, thanks! Please create a PR linked to https://github.com/thSoft/eclipse-commons/issues/61.

**Pull request merged**